lms/djangoapps/course_structure_api/v0/views.py

```diff
@@ -290,3 +296,299 @@ class CourseGradingPolicy(CourseViewMixin, ListAPIView):
+
+          * web_url: (string) The URL to the website location of this block.  This URL can be used as a further
+            fallback if the block_url and the block_json is not supported.
+
+        * navigation: A dictionary that maps block IDs to a collection of navigation information about each block.
+          Each block contains the following fields.
+
+          * descendants: (list) A list of IDs of the children of the block if the block's depth in the
+            course hierarchy is less than the navigation_depth.  Otherwise, a list of IDs of the aggregate descendants
+            of the block.
+
+        * blocks_navigation: A dictionary that combines both the blocks and navigation data.
+
+    """
+    @view_course_access(depth=None)
+    def list(self, request, course, return_blocks=True, return_nav=True, *args, **kwargs):
+
+        def json_field_requested(param_name, param_type, default_value):
+            """
+            Returns the JSON-parsed parameter value of the request's GET parameter of name, param_name.
+            If the parameter isn't specified in the request, returns the JSON parsing of the default_value.
+            """
+            field_requested = json.loads(request.GET.get(param_name, default_value))
+            if field_requested and not isinstance(field_requested, param_type):
+                raise ParseError
+            return field_requested
+
+        def bool_field_requested(param_name, default_value="true"):
+            """
+            Returns the boolean-parsed value of the request's GET parameter of name, param_name.
+            If the parameter isn't specified in the request, returns the parsing of the default_value.
+            The value is treated as True iff it's lowercase value is equal to "true".
+            """
+            field_requested = request.GET.get(param_name, default_value)
+            if field_requested and not isinstance(field_requested, basestring):
+                raise ParseError
+            return field_requested.lower() == "true"
+
+        # check what fields are requested
+        block_json_requested = json_field_requested('block_json', dict, "{}")
+        block_count_requested = json_field_requested('block_count', list, "[]")
+        graded_requested = bool_field_requested('graded')
+        responsive_ui_requested = bool_field_requested('responsive_ui')
+        children_requested = bool_field_requested('children')
+        navigation_depth_requested = int(request.GET.get('navigation_depth', '3'))
+
+        # prepare the response
+        response = {}
+        blocks = {}
+        navigation = {}
+        if return_blocks and return_nav:
+            navigation = blocks
+            response["blocks+navigation"] = blocks
+        elif return_blocks:
+            response["blocks"] = blocks
+        elif return_nav:
+            response["navigation"] = navigation
+
+        def recurse_blocks_nav(block, block_depth, descendants_of_parent):
```

Comment:
Does this really need to be a nested method? I've found that having very long nested methods makes code hard to read (I wish we didn't have them in the few other cases where we do).